### DISCUSSION TITLE
eCAL Recording on External HDD

### DISCUSSION THREAD

**Discussion by piyushmz** [eclipse-ecal/ecal#1020] [Q&A]
I've a camera topic which runs on 30hz with 6 cameras connected & 1920 x 1080 res. As the data is too heavy I cannot record it on NVME SSD, 44 Sec of record contains 32.7GB space on NVME. I've got one external HDD of 4TB [Transcend](https://us.transcend-info.com/product/external-hard-drive/storejet-25m3) connected to USB 3.2 Gen 1.
I've provided path of HDD to eCAL Recorder & Started recording but the after some time the recorder stops & doesn't record any data on HDD. The data gets corrupted & if try to open I get following.
`Unable to load measurement`
What am I doing wrong & what should I do to resolve it?

Thanks
Piyush

**Comment by KerstinKeller** [comment]
Hi @piyushmz,
in general nothing can be said against writing to an external HDD, but eCAL Recorder can record only as much data as your HDD will be able to write.

The way that eCAL Recorder works is that it buffers incoming messages and then dumps them to the HDD as fast as possible.
If the write performance is not good enough, it will allocate more and more memory, and at some point run out of it, depending on the available RAM on your system.

Please be aware, that a datarate of ~0,75GB / second (33GB / 44 seconds) is enormous. As per [Wikipedia](https://en.wikipedia.org/wiki/USB_3.0) USB 3.2 Gen 1 supports a theoretical write speed of 0.5 GB / s.
This is less than the 0.75 GB / s which you are trying to write. And this is only a theoretical speed. The hdf5 file format possibly adds some overhead, too.

So from the specs, you're producing more data that what can be written, and eCAL Recorder will run out of memory.
You could try with a faster HDD, or better even use an internal one.

Nevertheless, we should investigate and see that eCAL Recorder can shut down gracefully, and no data gets corrupted.

**Comment by piyushmz** [comment]
Hi @KerstinKeller so I've replaced my HDD with an external SSD & it works great but now further question is what if that 2TB SSD gets full? Where will ecal record data? Will it switch to internal SSD or just switch itself off.
I need some help with ecalsys. I've a program & it's running continuously & after some time it throws a segfault or an unwanted error & process dies. After the process dies I just need to capture this error so that I'll know the process have died & I can just send a trigger to a script. In simple way I want to capture this error & pass a trigger to a script(which will continue further instruction) that the process have died. How should I capture the error if process in ecal has died & trigger a script? Is there any possibility for doing this? Is yes what approach should I take to do this easily.
I've tried eCAL `Restart by severity` but I would use it later because it takes some time to restart the program.

**Reply by KerstinKeller** [reply]
No, if it runs out of disk space, unfortunately it will crash. We do have an issue for this.
About eCAL Sys, I cannot answer your question, maybe @FlorianReimold can.

**Reply by piyushmz** [reply]
Ok Thanks.

**Reply by FlorianReimold** [reply]
If your Task is an eCAL Task, you can use the "Restart by severity" feature to let it restart, if it crashes. But I would highly recommend you to fix your application, so it doesn't crash any more.

![grafik](https://user-images.githubusercontent.com/11774314/227544181-e7c11dec-7080-4c10-a2bb-6a6feacb17d8.png)

It will take some time for eCAL to realize your application has crashed.

**Reply by piyushmz** [reply]
@FlorianReimold Yes I've actually tired this. Thanks for your insight. What I was thinking that I create a bash script which will hold the PID's generated by ecalsys & check every few sec if the process has died? If yes continue else it'll trigger another script.
Is there any way that when I start ecal, I can get the PID of all the processes ecalsys have started & store them in an array & iterate over them to check if pid exists or no.

Thanks.

**Comment by FlorianReimold** [comment]
You can create such a script, however it will probably rather become a pyhton script than an bash script. You can use the `__ecalsys_state__` channel to get all the PIDs started by eCAL Sys and can from there act accordingly. But I would recommend using python for such a script, as eCAL has a python binding and protobuf is available for python as well.

You can see the `__ecalsys_state__` topic in the eCAL Monitor, if you check the "Show all" Checkbox:

![image](https://user-images.githubusercontent.com/11774314/227870331-0298e4ef-3020-47c6-9663-6652acff52bf.png)

You could also write a bash script that starts your faulty application in a while-true-loop. So when the old app terminates, it will immediately start a new instance. I have done that with nasty applications in the past and it worked sufficiently (no eCAL context). It will however prevent eCAL Sys from monitoring your application, as it will instead monitor the bash process running the script.

All in all it is still my recommendation to fix the faulty app instead of investing much work in nasty workarounds.

**Reply by piyushmz** [reply]
@FlorianReimold @KerstinKeller  Thanks for your help to resolve me this thing easily.
